File: lib/utilities/bloom_filter.dart
// Copyright (c) 2025 bjorge

import 'dart:convert';
import 'dart:typed_data';
import 'package:crypto/crypto.dart';

const defaultBloomFilterSize = 240;
const defaultBloomFilterHashCount = 7;

class BloomFilterProcessor {
  final int size; // Number of bits (m)
  final int hashCount; // Number of hash functions (k)
  final Uint8List bitArray;

  BloomFilterProcessor({
    required this.size,
    required this.hashCount,
    Uint8List? bitArray,
  }) : bitArray = bitArray ?? Uint8List((size / 8).ceil());

  void add(String item) {
    for (var i = 0; i < hashCount; i++) {
      final hash = _hash(item, i);
      final bitIndex = hash % size;
      _setBit(bitIndex);
    }
  }

  void addAll(List<String> items) {
    for (final item in items) {
      add(item);
    }
  }

  bool mightContain(String item) {
    for (var i = 0; i < hashCount; i++) {
      final hash = _hash(item, i);
      final bitIndex = hash % size;
      if (!_getBit(bitIndex)) return false;
    }
    return true; // Could be in the set (false positive possible)
  }

  void _setBit(int bitIndex) {
    final byteIndex = bitIndex ~/ 8;
    final bitMask = 1 << (bitIndex % 8);
    bitArray[byteIndex] |= bitMask;
  }

  bool _getBit(int bitIndex) {
    final byteIndex = bitIndex ~/ 8;
    final bitMask = 1 << (bitIndex % 8);
    return (bitArray[byteIndex] & bitMask) != 0;
  }

  int _hash(String input, int seed) {
    final bytes = utf8.encode('$seed:$input');
    final digest = sha256.convert(bytes);
    final byteList = Uint8List.fromList(digest.bytes);
    return byteList.buffer.asByteData().getUint32(0); // 32-bit hash
  }
}

File: packages/hyttahub_firebase_functions/src/auto_join/bloom_filter.ts
import * as crypto from "crypto";

export const defaultBloomFilterSize = 240;
export const defaultBloomFilterHashCount = 7;

export class BloomFilterProcessor {
  size: number;
  hashCount: number;
  bitArray: Uint8Array;

  constructor(size: number, hashCount: number, bitArray?: Uint8Array) {
    this.size = size;
    this.hashCount = hashCount;
    this.bitArray = bitArray || new Uint8Array(Math.ceil(size / 8));
  }

  add(item: string) {
    for (let i = 0; i < this.hashCount; i++) {
      const hash = this._hash(item, i);
      const bitIndex = hash % this.size;
      this._setBit(bitIndex);
    }
  }

  addAll(items: string[]) {
    for (const item of items) {
      this.add(item);
    }
  }

  private _setBit(bitIndex: number) {
    const byteIndex = Math.floor(bitIndex / 8);
    const bitMask = 1 << (bitIndex % 8);
    this.bitArray[byteIndex] |= bitMask;
  }

  mightContain(item: string): boolean {
    for (let i = 0; i < this.hashCount; i++) {
      const hash = this._hash(item, i);
      const bitIndex = hash % this.size;
      const byteIndex = Math.floor(bitIndex / 8);
      const bitMask = 1 << (bitIndex % 8);
      if ((this.bitArray[byteIndex] & bitMask) === 0) {
        return false;
      }
    }
    return true;
  }

  private _hash(input: string, seed: number): number {
    const h = crypto.createHash("sha256");
    h.update(`${seed}:${input}`);
    const digest = h.digest();
    // Match Dart's ByteData.getUint32(0) which is Big-endian
    return digest.readUInt32BE(0);
  }
}
